### PR TITLE
[2.x] Allows to cache job attempts on cache

### PR DESCRIPTION
**Pull request by nunomaduro**
This pull request improves Vapor's way of handling job attempts.

Check more info here: https://github.com/laravel/vapor/pull/347

### CODE REVIEW

File: src/Queue/JobAttempts.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    /**
+     * Transfer the job attemps.
+     *
+     * @param  \Laravel\Vapor\Queue\VaporJob|string  $from
+     * @param  \Laravel\Vapor\Queue\VaporJob|string  $to
+     * @return void
+     */
+    public function transfer($from, $to)
+    {
+        $this->cache->put($this->key($to), $this->get($from));
```
Comment (themsaid) [inline src/Queue/JobAttempts.php]: I think you missed the $seconds argument of the put() method.

File: src/Queue/JobAttempts.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * Increment the number of times the job has been attempted.
+     *
+     * @param  \Laravel\Vapor\Queue\VaporJob  $job
+     * @return void
+     */
+    public function increment($job)
+    {
+        $this->cache->increment($this->key($job));
```
Comment (themsaid) [inline src/Queue/JobAttempts.php]: I think you cannot increment a key the doesn't exist yet in DynamoDB. Can you verify that?

Comment (nunomaduro) [reply src/Queue/JobAttempts.php]: Fixed.

File: src/Queue/JobAttempts.php (hunk 1 above)
Comment (nunomaduro) [reply src/Queue/JobAttempts.php]: Fixed.

**Comment by nunomaduro** [comment]
Tested on Vapor, and things work as expected. The only thing that is not clear for me, is the costs by using the cache on every queue job.

**Comment by jerrebm** [comment]
@nunomaduro It looks like this update to the `vapor-core` package has caused a significant increase in our DynamoDB read/write requests and thereby a **40x increase of our DynamoDB costs**. We were spending ~$0.50 per day, which instantly jumped to $20+ per day:

cost-increase

It took me a while to figure out the cause of this sudden spike as we did not make any changes to our code that would have increased calls to DynamoDB (we barely use the Cache at all), nor has there been a significant change in the number of jobs being processed or traffic handled.

Through the AWS console I eventually found that a massive amount of read/write requests to Dynamo were being made for `laravel_vapor_job_attemps:*` keys. Rolling back `vapor-core` to `v2.20.0`, before this change was implemented, dropped read/writes to Dynamo back to normal:

read-write-stats

Do we have the option to disable this feature so we can avoid these increased costs and requests?

**Comment by nunomaduro** [comment]
@jerrebm Joe have answered you via the Vapor support channel.

**Comment by jerrebm** [comment]
@nunomaduro Thank you, well received! I think it's worth considering adding a big warning in the documentation about this?

I'll add Joe's reply below in case anyone runs into this problem in the future:

> You may disable this safeguard by setting "queue-cache-job-attempts: false" in your vapor.yml file. However, please be advised that if your job is timed out by Lambda before it has completed or if you inadvertantly leave a dd() in your handler, it will continue to run indefinitely which can lead to unexpected increase in charges for services such as CloudWatch or Parameter Store.
>
>Instead, you may find switching to a Redis cache may resolve both issues.

I'll try switching to Redis. Thank you both for the quick reply! 😃

**Pull request merged**